### ISSUE TITLE
missing "New Caddyfile" creation

### ISSUE THREAD

**Issue by devhero** [lucaslorentz/caddy-docker-proxy#409]
I made work the basic example https://github.com/lucaslorentz/caddy-docker-proxy#basic-usage-example-using-docker-compose in my linux laptop (Fedora 36), otherwise under a Linode VM (Ubuntu 22.04 LTS) it doesn't work, namely "whoami" container can not be reached.
After a lot of attempts I figured out what's wrong watching at the logs.

In the not working env (Linode VM) the logs are the following:

```shell
caddy-caddy-1  | {"level":"info","ts":1664975003.4609492,"logger":"docker-proxy","msg":"Running caddy proxy server"}
caddy-caddy-1  | {"level":"info","ts":1664975003.4622898,"logger":"admin","msg":"admin endpoint started","address":"localhost:2019","enforce_origin":false,"origins":["//localhost:2019","//[::1]:2019","//127.0.0.1:2019"]}
caddy-caddy-1  | {"level":"info","ts":1664975003.4627542,"msg":"autosaved config (load with --resume flag)","file":"/config/caddy/autosave.json"}
caddy-caddy-1  | {"level":"info","ts":1664975003.4629436,"logger":"docker-proxy","msg":"Running caddy proxy controller"}
caddy-caddy-1  | {"level":"info","ts":1664975003.4742067,"logger":"docker-proxy","msg":"Start","CaddyfilePath":"","LabelPrefix":"caddy","PollingInterval":30,"ProcessCaddyfile":true,"ProxyServiceTasks":true,"IngressNetworks":"[caddy]","DockerSockets":[""],"DockerCertsPath":[""],"DockerAPIsVersion":[""]}
caddy-caddy-1  | {"level":"info","ts":1664975003.4766765,"logger":"docker-proxy","msg":"IngressNetworksMap","ingres":"map[]"}
caddy-caddy-1  | {"level":"info","ts":1664975003.490815,"logger":"docker-proxy","msg":"Connecting to docker events","DockerSocket":""}
caddy-caddy-1  | {"level":"info","ts":1664975003.4921694,"logger":"docker-proxy","msg":"Swarm is available","new":false}
caddy-caddy-1  | {"level":"info","ts":1664975003.4927227,"logger":"docker-proxy","msg":"New Caddyfile","caddyfile":"# Empty caddyfile"}
caddy-caddy-1  | {"level":"warn","ts":1664975003.4931984,"logger":"docker-proxy","msg":"Caddyfile to json warning","warn":"[Caddyfile:1: Caddyfile input is not formatted; run the 'caddy fmt' command to fix inconsistencies]"}
caddy-caddy-1  | {"level":"info","ts":1664975003.4934154,"logger":"docker-proxy","msg":"New Config JSON","json":"{}"}
caddy-caddy-1  | {"level":"info","ts":1664975003.4936678,"logger":"docker-proxy","msg":"Sending configuration to","server":"localhost"}
caddy-caddy-1  | {"level":"info","ts":1664975003.4944704,"logger":"admin.api","msg":"received request","method":"POST","host":"localhost:2019","uri":"/load","remote_ip":"127.0.0.1","remote_port":"40426","headers":{"Accept-Encoding":["gzip"],"Content-Length":["41"],"Content-Type":["application/json"],"User-Agent":["Go-http-client/1.1"]}}
caddy-caddy-1  | {"level":"info","ts":1664975003.4947662,"msg":"config is unchanged"}
caddy-caddy-1  | {"level":"info","ts":1664975003.495951,"logger":"admin.api","msg":"load complete"}
caddy-caddy-1  | {"level":"info","ts":1664975003.496327,"logger":"docker-proxy","msg":"Successfully configured","server":"localhost"}
```

while in my laptop logs are:

```shell
caddy-caddy-1  | {"level":"info","ts":1664975090.7126575,"logger":"docker-proxy","msg":"Running caddy proxy server"}
caddy-caddy-1  | {"level":"info","ts":1664975090.714367,"logger":"admin","msg":"admin endpoint started","address":"localhost:2019","enforce_origin":false,"origins":["//127.0.0.1:2019","//localhost:2019","//[::1]:2019"]}
caddy-caddy-1  | {"level":"info","ts":1664975090.7145946,"msg":"autosaved config (load with --resume flag)","file":"/config/caddy/autosave.json"}
caddy-caddy-1  | {"level":"info","ts":1664975090.714606,"logger":"docker-proxy","msg":"Running caddy proxy controller"}
caddy-caddy-1  | {"level":"info","ts":1664975090.7157886,"logger":"docker-proxy","msg":"Start","CaddyfilePath":"","LabelPrefix":"caddy","PollingInterval":30,"ProcessCaddyfile":true,"ProxyServiceTasks":true,"IngressNetworks":"[caddy]","DockerSockets":[""],"DockerCertsPath":[""],"DockerAPIsVersion":[""]}
caddy-caddy-1  | {"level":"info","ts":1664975090.7200892,"logger":"docker-proxy","msg":"Connecting to docker events","DockerSocket":""}
caddy-caddy-1  | {"level":"info","ts":1664975090.7207496,"logger":"docker-proxy","msg":"IngressNetworksMap","ingres":"map[8408cf3333210c72dceedfcf21726bb9537a428354618773398a4592e4948d2b:true]"}
caddy-caddy-1  | {"level":"info","ts":1664975090.7589445,"logger":"docker-proxy","msg":"Swarm is available","new":true}
caddy-caddy-1  | {"level":"info","ts":1664975090.766748,"logger":"docker-proxy","msg":"New Caddyfile","caddyfile":"# Empty caddyfile"}
caddy-caddy-1  | {"level":"warn","ts":1664975090.767255,"logger":"docker-proxy","msg":"Caddyfile to json warning","warn":"[Caddyfile:1: Caddyfile input is not formatted; run the 'caddy fmt' command to fix inconsistencies]"}
caddy-caddy-1  | {"level":"info","ts":1664975090.7673006,"logger":"docker-proxy","msg":"New Config JSON","json":"{}"}
caddy-caddy-1  | {"level":"info","ts":1664975090.7673519,"logger":"docker-proxy","msg":"Sending configuration to","server":"localhost"}
caddy-caddy-1  | {"level":"info","ts":1664975090.771526,"logger":"admin.api","msg":"received request","method":"POST","host":"localhost:2019","uri":"/load","remote_ip":"127.0.0.1","remote_port":"59454","headers":{"Accept-Encoding":["gzip"],"Content-Length":["41"],"Content-Type":["application/json"],"User-Agent":["Go-http-client/1.1"]}}
caddy-caddy-1  | {"level":"info","ts":1664975090.7716184,"msg":"config is unchanged"}
caddy-caddy-1  | {"level":"info","ts":1664975090.7716603,"logger":"admin.api","msg":"load complete"}
caddy-caddy-1  | {"level":"info","ts":1664975090.7718525,"logger":"docker-proxy","msg":"Successfully configured","server":"localhost"}

### so far it's the same

caddy-caddy-1  | {"level":"info","ts":1664975094.3147235,"logger":"docker-proxy","msg":"New Caddyfile","caddyfile":"localhost {\n\treverse_proxy 172.24.0.3:8000\n}\n"}
caddy-caddy-1  | {"level":"info","ts":1664975094.3152227,"logger":"docker-proxy","msg":"New Config JSON","json":"{\"apps\":{\"http\":{\"servers\":{\"srv0\":{\"listen\":[\":443\"],\"routes\":[{\"match\":[{\"host\":[\"localhost\"]}],\"handle\":[{\"handler\":\"subroute\",\"routes\":[{\"handle\":[{\"handler\":\"reverse_proxy\",\"upstreams\":[{\"dial\":\"172.24.0.3:8000\"}]}]}]}],\"terminal\":true}]}}}}}"}
caddy-caddy-1  | {"level":"info","ts":1664975094.3152695,"logger":"docker-proxy","msg":"Sending configuration to","server":"localhost"}
caddy-caddy-1  | {"level":"info","ts":1664975094.3154762,"logger":"admin.api","msg":"received request","method":"POST","host":"localhost:2019","uri":"/load","remote_ip":"127.0.0.1","remote_port":"59454","headers":{"Accept-Encoding":["gzip"],"Content-Length":["287"],"Content-Type":["application/json"],"User-Agent":["Go-http-client/1.1"]}}
caddy-caddy-1  | {"level":"info","ts":1664975094.3163345,"logger":"admin","msg":"admin endpoint started","address":"localhost:2019","enforce_origin":false,"origins":["//127.0.0.1:2019","//localhost:2019","//[::1]:2019"]}
caddy-caddy-1  | {"level":"info","ts":1664975094.3165138,"logger":"http","msg":"server is listening only on the HTTPS port but has no TLS connection policies; adding one to enable TLS","server_name":"srv0","https_port":443}
caddy-caddy-1  | {"level":"info","ts":1664975094.3165295,"logger":"http","msg":"enabling automatic HTTP->HTTPS redirects","server_name":"srv0"}
caddy-caddy-1  | {"level":"info","ts":1664975094.318018,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/data/caddy"}
caddy-caddy-1  | {"level":"info","ts":1664975094.318345,"logger":"tls","msg":"finished cleaning storage units"}
caddy-caddy-1  | {"level":"info","ts":1664975094.3185122,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc0001dae00"}
caddy-caddy-1  | {"level":"info","ts":1664975094.3457537,"logger":"pki.ca.local","msg":"root certificate is already trusted by system","path":"storage:pki/authorities/local/root.crt"}
caddy-caddy-1  | {"level":"info","ts":1664975094.3461013,"logger":"http","msg":"enabling HTTP/3 listener","addr":":443"}
caddy-caddy-1  | {"level":"info","ts":1664975094.3463209,"msg":"failed to sufficiently increase receive buffer size (was: 208 kiB, wanted: 2048 kiB, got: 416 kiB). See https://github.com/lucas-clemente/quic-go/wiki/UDP-Receive-Buffer-Size for details."}
caddy-caddy-1  | {"level":"info","ts":1664975094.3464663,"logger":"http.log","msg":"server running","name":"srv0","protocols":["h1","h2","h3"]}
caddy-caddy-1  | {"level":"info","ts":1664975094.346554,"logger":"http.log","msg":"server running","name":"remaining_auto_https_redirects","protocols":["h1","h2","h3"]}
caddy-caddy-1  | {"level":"info","ts":1664975094.3465817,"logger":"http","msg":"enabling automatic TLS certificate management","domains":["localhost"]}
caddy-caddy-1  | {"level":"warn","ts":1664975094.3470113,"logger":"tls","msg":"stapling OCSP","error":"no OCSP stapling for [localhost]: no OCSP server specified in certificate","identifiers":["localhost"]}
caddy-caddy-1  | {"level":"info","ts":1664975094.3471894,"msg":"autosaved config (load with --resume flag)","file":"/config/caddy/autosave.json"}
caddy-caddy-1  | {"level":"info","ts":1664975094.3472252,"logger":"admin.api","msg":"load complete"}
caddy-caddy-1  | {"level":"info","ts":1664975094.347396,"logger":"docker-proxy","msg":"Successfully configured","server":"localhost"}
caddy-caddy-1  | {"level":"info","ts":1664975094.3610954,"logger":"admin","msg":"stopped previous server","address":"localhost:2019"}
```

In my laptop I restarted the whoami container and I've seen that every time it goes down/up, DCP recreates the Caddyfile.
Instead, under Linode VM, CDP can't listen the shutdown/startup of whoami container and never creates the Caddyfile.

Linode VM shell:
```bash
$ docker ps
CONTAINER ID   IMAGE                                       COMMAND                  CREATED             STATUS          PORTS                                                                                NAMES
fed4fc1b4789   jwilder/whoami                              "/app/http"              About an hour ago   Up 9 seconds    8000/tcp                                                                             whoami-whoami-1
8e4a90d5c0b0   lucaslorentz/caddy-docker-proxy:ci-alpine   "/bin/caddy docker-p…"   About an hour ago   Up 39 seconds   0.0.0.0:80->80/tcp, :::80->80/tcp, 0.0.0.0:443->443/tcp, :::443->443/tcp, 2019/tcp   caddy-caddy-1
```

I can't understand what is missing to make CDP listen other containers under the same network and the same machine.

**Comment by devhero** [comment]
After a lot of other tests, I fixed the issue, which I'm quite sure was the docker's [rootless mode](https://docs.docker.com/engine/security/rootless/).
I repeated step-by-step the procedure and maybe I've missed some configuration which didn't allow the non-root user to make communicating containers each others.

**Comment by devhero** [comment]
;-)

**Comment by lucaslorentz** [comment]
This line shows CDP didn't recognize any network connected to it.
```
caddy-caddy-1  | {"level":"info","ts":1664975003.4766765,"logger":"docker-proxy","msg":"IngressNetworksMap","ingres":"map[]"}
```
IngressNetworksMap should have the id of the networks CDP will use to connect to other containers.

**Comment by devhero** [comment]
Thanks for the quick response!
I already made the set of CADDY_INGRESS_NETWORKS, as it's included in the basic example.
I suppose that issues from docker rootless wrong settings blocked some mechanism that allow network recognition.
In fact I had similar issues with Traefik, which didn't recognize docker containers trying to connect to it.
For this reason I switched to Caddy, but now that I'm in with Caddy, I feel very comfortable :smiley:
